includes/internal/quiz-submission/submission/repositories/class-aggregate-submission-repository.php

```diff
@@ -0,0 +1,171 @@
+	 */
+	public function __construct( Comments_Based_Submission_Repository $comments_based_repository, Tables_Based_Submission_Repository $tables_based_repository, bool $use_tables ) {
+		$this->comments_based_repository = $comments_based_repository;
+		$this->tables_based_repository   = $tables_based_repository;
+		$this->use_tables                = $use_tables;
+	}
+
+	/**
+	 * Creates a new quiz submission.
+	 *
+	 * @internal
+	 *
+	 * @param int        $quiz_id     The quiz ID.
+	 * @param int        $user_id     The user ID.
+	 * @param float|null $final_grade The final grade.
+	 *
+	 * @return Submission The quiz submission.
+	 */
+	public function create( int $quiz_id, int $user_id, float $final_grade = null ): Submission {
+		$submission = $this->comments_based_repository->create( $quiz_id, $user_id, $final_grade );
+
+		if ( $this->use_tables ) {
+			$this->tables_based_repository->create( $quiz_id, $user_id, $final_grade );
+		}
+
+		return $submission;
+	}
+
+	/**
+	 * Get or create a new quiz submission if it doesn't exist.
+	 *
+	 * @internal
+	 *
+	 * @param int        $quiz_id     The quiz ID.
+	 * @param int        $user_id     The user ID.
+	 * @param float|null $final_grade The final grade.
+	 *
+	 * @return Submission The quiz submission.
+	 */
+	public function get_or_create( int $quiz_id, int $user_id, float $final_grade = null ): Submission {
+		$submission = $this->get( $quiz_id, $user_id );
+
+		if ( $submission ) {
+			return $submission;
+		}
+
+		return $this->create( $quiz_id, $user_id, $final_grade );
+	}
+
+	/**
+	 * Gets a quiz submission.
+	 *
+	 * @internal
+	 *
+	 * @param int $quiz_id The quiz ID.
+	 * @param int $user_id The user ID.
+	 *
+	 * @return Submission|null The quiz submission.
+	 */
+	public function get( int $quiz_id, int $user_id ): ?Submission {
+		return $this->comments_based_repository->get( $quiz_id, $user_id );
+	}
+
+	/**
+	 * Get the questions related to the quiz submission.
+	 *
+	 * @internal
+	 *
+	 * @param int $submission_id The quiz submission ID.
+	 *
+	 * @return array An array of question post IDs.
+	 */
+	public function get_question_ids( int $submission_id ): array {
+		return $this->comments_based_repository->get_question_ids( $submission_id );
+	}
+
+	/**
+	 * Save quiz submission.
+	 *
+	 * @internal
+	 *
+	 * @param Submission $submission The quiz submission.
+	 */
+	public function save( Submission $submission ): void {
+		$this->comments_based_repository->save( $submission );
+
+		if ( $this->use_tables ) {
+			$tables_based_submission = $this->tables_based_repository->get( $submission->get_quiz_id(), $submission->get_user_id() );
+
+			if ( $tables_based_submission ) {
+				$submission_to_save = new Submission(
+					$tables_based_submission->get_id(),
+					$submission->get_quiz_id(),
+					$submission->get_user_id(),
+					$submission->get_final_grade(),
+					$submission->get_created_at(),
+					$submission->get_updated_at()
```

Comment:
Do you need to convert it to UTC? ([Like I did here.](https://github.com/Automattic/sensei/pull/5894/commits/37308da8a7cffa315ff53696d8c73acdd5e8d043))

Reply:
I don't have the issue you were facing so I'm not sure.

I've noticed that you are using the WordPress timezone [when creating the course progress model](https://github.com/Automattic/sensei/blob/5e54866546e5aae8d4f4ec5af87e0216494cb37a/includes/internal/student-progress/course-progress/repositories/class-comments-based-course-progress-repository.php#L74) but UTC [when creating the lesson progress model](https://github.com/Automattic/sensei/blob/55c6c7feae6f22734ec1a6856ea69f607b6180d9/includes/internal/student-progress/lesson-progress/repositories/class-tables-based-lesson-progress-repository.php#L123). Does it make sense to always use UTC in the models to avoid potential date issues? What do you think of forcing the datetime variables in the models to always be in UTC? Can you think of another solution that can help with this?

Reply:
Hmm, the problem in my PR was because we create the comments-based progress with WP TZ, but we need to convert it to UTC, when assign it to the tables-based progress.

I found you create the comment-based submission with internal TZ too. But I don't see the conversion while saving in the aggregate repository. Now I see, you force set updated at, so indeed you don't have that bug :) (However, that temporary object has wrong dates.)

(By the way, I'm not sure it's good to force updated_at here as it might cause different timestamps for their two versions. That's not critical probably if we're not going to use them during sync/migration.)

Reply:
> I've noticed that you are using the WordPress timezone [when creating the course progress model](https://github.com/Automattic/sensei/blob/5e54866546e5aae8d4f4ec5af87e0216494cb37a/includes/internal/student-progress/course-progress/repositories/class-comments-based-course-progress-repository.php#L74) but UTC [when creating the lesson progress model](https://github.com/Automattic/sensei/blob/55c6c7feae6f22734ec1a6856ea69f607b6180d9/includes/internal/student-progress/lesson-progress/repositories/class-tables-based-lesson-progress-repository.php#L123).

It is exactly what I described:
- first link refers to comments-based implementation ([and UTC for tables-based](https://github.com/Automattic/sensei/blob/55c6c7feae6f22734ec1a6856ea69f607b6180d9/includes/internal/student-progress/course-progress/repositories/class-tables-based-course-progress-repository.php#L55))
- second link refers to tables-based implementation ([and WP TZ for comments-based](https://github.com/Automattic/sensei/blob/5e54866546e5aae8d4f4ec5af87e0216494cb37a/includes/internal/student-progress/lesson-progress/repositories/class-comments-based-lesson-progress-repository.php#L40))

Reply:
> I found you create the comment-based submission with internal TZ too.

Oh, right! This might turn into an issue when we start doing the migrations.

Do you think we can do something about this complexity? To me, it's weird to use one timezone when creating the model from the comments repository and another when using the tables. Maybe we switch from using WP timezone when working with comments to UTC?

Reply:
That's hard. I don't think it is a good idea to use UTC as that way we change the existing behaviour (it leads to having unexpected dates in DB).

Maybe we just have to pay more attention now, and after removing comments-based logic it won't make sense anymore.

Reply:
OK, that makes sense. 👍 🙂 

I will do a repository update in a separate PR to account for dates that may not be in UTC.